store/tikv/region_request.go

```diff
@@ -55,109 +55,384 @@ func NewRegionRequestSender(bo *Backoffer, regionCache *RegionCache, client Clie
 	}
 }
 
-// SendKVReq sends a KV request to tikv server.
-func (s *RegionRequestSender) SendKVReq(req *kvrpcpb.Request, regionID RegionVerID, timeout time.Duration) (*kvrpcpb.Response, error) {
+// RegionErrorResponse is an interface used for retry on region errors in
+// the implemetation of RegionRequestSender.
+type RegionErrorResponse interface {
+	GetRegionError() *errorpb.Error
+}
+
+type rpcFunc func(ctx *RPCContext) (RegionErrorResponse, error, bool)
+
+func (s *RegionRequestSender) callRPCFunc(regionID RegionVerID, f rpcFunc) error {
 	for {
 		ctx, err := s.regionCache.GetRPCContext(s.bo, regionID)
 		if err != nil {
-			return nil, errors.Trace(err)
-		}
-		if ctx == nil {
-			// If the region is not found in cache, it must be out
-			// of date and already be cleaned up. We can skip the
-			// RPC by returning RegionError directly.
-			return &kvrpcpb.Response{
-				Type:        req.GetType(),
-				RegionError: &errorpb.Error{StaleEpoch: &errorpb.StaleEpoch{}},
-			}, nil
+			return errors.Trace(err)
 		}
-
-		resp, retry, err := s.sendKVReqToRegion(ctx, req, timeout)
+		resp, err, retry := f(ctx)
 		if err != nil {
-			return nil, errors.Trace(err)
-		}
-		if retry {
+			if e := s.onSendFail(ctx, err); e != nil {
+				return errors.Trace(e)
+			}
 			continue
 		}
 
+		if !retry {
+			return nil
+		}
+
 		if regionErr := resp.GetRegionError(); regionErr != nil {
 			retry, err := s.onRegionError(ctx, regionErr)
 			if err != nil {
-				return nil, errors.Trace(err)
+				return errors.Trace(err)
 			}
 			if retry {
 				continue
 			}
-			return resp, nil
+			return nil
 		}
 
-		if resp.GetType() != req.GetType() {
-			return nil, errors.Trace(errMismatch(resp, req))
-		}
-		return resp, nil
+		return nil
 	}
 }
 
-// SendCopReq sends a coprocessor request to tikv server.
-func (s *RegionRequestSender) SendCopReq(req *coprocessor.Request, regionID RegionVerID, timeout time.Duration) (*coprocessor.Response, error) {
-	for {
-		ctx, err := s.regionCache.GetRPCContext(s.bo, regionID)
-		if err != nil {
-			return nil, errors.Trace(err)
-		}
+// KvGet sends a Get request to TiKV server.
+func (s *RegionRequestSender) KvGet(req *kvrpcpb.GetRequest, regionID RegionVerID, timeout time.Duration) (*kvrpcpb.GetResponse, error) {
+	var resp *kvrpcpb.GetResponse
+	f := func(ctx *RPCContext) (RegionErrorResponse, error, bool) {
 		if ctx == nil {
 			// If the region is not found in cache, it must be out
 			// of date and already be cleaned up. We can skip the
 			// RPC by returning RegionError directly.
-			return &coprocessor.Response{
+			resp = &kvrpcpb.GetResponse{
```

Comment:
Let's add a TODO and postpone this change. This PR is huge already, it's not proper to add more complexity to it.